feat: Krist WebSocket refactor

bananasov opened the pull request:
This PR aims to rewrite the majority of our WebSocket handling, this includes the initial connection and the communication between server and client.

So far, broadcasting has been implemented to send events back to a connected client, event subscriptions are not handled yet.

bananasov committed:
chore: Remove WsDataManager struct

This struct was never used in the end, now it's gone

bananasov committed:
feat: Add needed libraries for the WebSocket refactor

dashmap is a faster hashmap implementation, bytestring is used for communication

bananasov committed:
chore: Remove leftovers of WsDataManager

bananasov committed:
feat: Implement new WebSocket server for refactor

bananasov committed:
feat: Rewrite the `/ws/start` route to use new WebSocket server

bananasov committed:
chore: Remove more leftovers from old WsDataManager and TokenCache

bananasov committed:
feat: Implement WebSocket server related errors

bananasov committed:
feat: Implement initial handling of connection for `/gateway/{token}`

bananasov committed:
fix: Make session insertion and deletion functions public

bananasov committed:
feat: Add heartbeat and message handling template for WebSocket refactor

bananasov committed:
fix: Actually use web::Data for app data call

bananasov committed:
feat: Implement Keepalive packet

bananasov committed:
chore: Remove TokenCache and old WsServer implementation

bananasov committed:
feat: Implement message handling

currently this is just bolted on, I have plans to rewrite all of the route stuff later but not now at 2 am.

bananasov commented:
@uerkw So far so good, I've got MSKS and Radon connecting without issues, one slight thing I've noticed though is when you exit a Minecraft world, websockets don't get closed (and I'm unsure if we handle dropped connections correctly in this impl or the original one)

Furthermore, I am planning on extracting the functionality of the Websocket server into a trait so we can more easily run a krist websocket impl and a kromer one. I am thinking of rewriting the route stuff as well as it's a bit "bloaty" but that'll come in a different PR.

bananasov committed:
chore: Remove unused dependency `actix`

bananasov committed:
fix: Don't do `.to_string()` when converting a transaction into its krist counterpart

bananasov committed:
feat: Implement `broadcast_event` function

bananasov committed:
chore: Cleanup some code related to address session modification

bananasov committed:
feat: Implement transaction event for websockets

NOTE: We are not doing ANY checks yet related to which subscriptions a session has, this should be fixed.

bananasov committed:
chore: Remove WrappedWsData and its references

bananasov committed:
feat: Implement new functions for managing subscriptions

bananasov committed:
chore: Remove more WrappedWsData references

bananasov committed:
chore: Remove `WsSessionModification` and its references

this also includes commenting out all of the existing routes, rewriting them later

bananasov committed:
feat: Begin rewriting WebSocket message handling

bananasov committed:
chore: Rewrite WebSocket message parsing to be more concise (subjective)

ElonGaties commented on src/routes/krist/ws.rs:
```diff
@@ -1,23 +1,24 @@
 use std::str::FromStr;
-
-use actix::spawn;
-//use actix::prelude::*;
```
i honestly dont like prelude includes, based

bananasov replied:
I didn't even notice i removed those tbh

bananasov committed:
feat: More rewriting of message handling

bananasov committed:
feat: Add functions for subscribing and unsubscribing to an event

Also adds a function to get all subscribed events for a session (get_subscription_list)

ElonGaties commented on src/routes/krist/ws.rs:
```diff
@@ -90,89 +68,156 @@ pub async fn setup_ws(
-        return send_error_message(req.clone(), body).await;
+        return Ok(response);
     }
+    let uuid = uuid_result.unwrap(); // SAFETY: We handled the error above
```
might swap this out to be a match statement, rather than an if - purely to avoid the unwrap

bananasov committed:
feat: Add a note about the potational erroring in message IDs

Sometimes krist shops send *numbers* as a *string*, fucking hell.

bananasov committed:
chore: Comment out some code that errors (broadcasting transactions)

bananasov commented:
I cast 10 more commits upon ye.

So this amalgamation of commits got the majority of stuff working, that includes subscriptions, logins and logouts. I might've rewritten all of the websocket message parsing while i was programming like a crackhead with a computer science degree

bananasov committed:
feat: Implement `me` route for WebSockets

bananasov committed:
feat: Implement `address` route for WebSockets

bananasov committed:
feat: Implement `make_transaction` route for WebSockets

bananasov committed:
feat: Implement `work` and default route for WebSockets

bananasov commented:
And that should be all!

bananasov committed:
feat: Implement WebSocket event messages

and that marks the full refactor of the WebSocket system

bananasov commented:
We're now missing one last function and that is subscription checks for sessions, this shouldn't be too hard to implement I reckon.

Commit 5ea0c32be88a1d5f54c32cc28fa5aa84240943bc marks the full refactor of our WebSocket system, I plan to make this more modular for v3 but that will come later

uerkw committed:
wip: openapi generation test

a55uka committed:
fix: use match statements over if for handling

a55uka committed:
fix: make from and to non-optional as they will always be populated

a55uka committed:
fix: make NameJson fit Name model

a55uka committed:
feat: broadcast Name and Transaction events to subscribers

a55uka committed:
clean: update lock and cargo clippy and fmt

a55uka committed:
clean: upgrade packages

a55uka committed:
clean: update pnpm pkgs

bananasov committed:
fix: Fix not being able to log in from WebSockets

bananasov committed:
fix: Don't clone sessions, not needed

bananasov committed:
fix: DO NOT CALL BROADCAST IN BROADCAST_EVENT

AAAAAA

bananasov committed:
feat: Return an error when invalid an event is invalid

bananasov committed:
fix: Cleanup session if it is closed

uerkw committed:
fix: name wallet for name transactions. Should be unaccessible. Will set to burn wallet balance

a55uka committed:
feat: monitor db connection, close when dead

a55uka committed:
feat: get session data from internal api

a55uka committed:
fix: unneeded namespace and generated surreal migration

a55uka committed:
clean: cargo fmt

a55uka committed:
fix: loop through protein instead of hard coded number

a55uka committed:
fix: use serde feature from dash* crate

a55uka committed:
fix: remove unused struct and imports

bananasov committed:
fix: Cleanup session when it closed due to inactivity

bananasov committed:
fix: Send proper error response for unscubscrib route

We weren't sending a message for when it catches an invalid event when unsubscribing, whoopsie

uerkw committed:
feat: a basic route schema for docs

bananasov committed:
fix: Make `FORCE_WS_INSECURE` optional

bananasov committed:
chore: Upgrade dependencies

serde: 1.0.217 => 1.0.218
serde_json: 1.0.138 => 1.0.139
surrealdb: 2.2.0 => 2.2.1

bananasov committed:
feat: Rename `x` parameter to `record` in serializer functions

bananasov committed:
feat: Add record ID serializers for JSON models

EmmaKnijn commented:
lgtm, will merge when @dimaguy approves

dimaguy commented:
> lgtm, will merge when @dimaguy approves

This codebase is not up to us :p
